### DISCUSSION TITLE
Import error while using in a Traefik plugin

### DISCUSSION THREAD

**Discussion by roman-yerin** [rabbitmq/amqp091-go#235] [Q&A]
### Describe the bug

I am trying to implement a Traefik plugin here https://github.com/ion-toolbox/traefik-jwt-eddsa
When Traefik loads the plugin, I receive an error message

`ERR github.com/traefik/traefik/v3/cmd/traefik/traefik.go:220 > Plugins are disabled because an error has occurred. error="github.com/ion-toolbox/traefik-jwt-eddsa: failed to import plugin code \"github.com/ion-toolbox/traefik-jwt-eddsa\": 1:21: import \"github.com/ion-toolbox/traefik-jwt-eddsa\" error: plugins-local/src/github.com/ion-toolbox/traefik-jwt-eddsa/plugin.go:9:2: import \"github.com/rabbitmq/amqp091-go\" error: plugins-local/src/github.com/ion-toolbox/traefik-jwt-eddsa/vendor/github.com/rabbitmq/amqp091-go/channel.go:425:18: cannot use type func(amqp091.frame) as type func(*struct {},amqp091.frame)"
`

As you can see, an exception fires at the import directive, i.e. my own code doesn't start.
FYI: as per traefik plugin documentation, all the modules have to be vendored.

### Reproduction steps

1. Checkout the repository https://github.com/ion-toolbox/traefik-jwt-eddsa
2. docker compose up

### Expected behavior

The plugin should start

**Comment by Zerpet** [comment]
I've created a 7 line program that creates a connection, a channel, and closes them both, vendor'ed the modules and it builds fine. I highly doubt that vendoring this module is broken in any way, and it looks like something specific to your environment/setup/code.

<details>
<summary>go.mod</summary>

```go
module localtest

go 1.21.5

require github.com/rabbitmq/amqp091-go v1.9.0 // indirect
```
</details>

<details>
<summary>main.go</summary>

```go
package main

import (
	"log"

	amqp "github.com/rabbitmq/amqp091-go"
)

func main() {
  log.Println("hello world!")
  conn, _ := amqp.Dial("amqp://localhost")
  ch, _ := conn.Channel()
  log.Println("I have a channel!") 
  conn.Close()
  ch.Close()
  log.Println("Bye world!")
}
```
</details>

<details>
<summary>commands</summary>

```zsh
go get github.com/rabbitmq/amqp091-go
go mod vendor
go build -mod=vendor main.go
```
</details>

I'll be happy to look at this issue if you can repro without traefik, or other external dependencies, in the mix.

**Comment by roman-yerin** [comment]
I never said that the problem was related to the placement of the module in the vendor directory, I just mentioned all the details.
The problem occurs in the traefik plugin environment, so I don't think reproducing it without traefik makes any sense. Other modules don't pose any import issues and as you can see the problem is in channel.go:425 so I think this repository is the right place for this issue. I don't think the problem has anything to do with my code because the plugin crashes before it even runs. This is either a library bug or a problem loading the traefik plugin, and it would be great to start here before the traefik repository.

**Comment by roman-yerin** [comment]
According to my research, this issue may be related to Yaegi, a traefik plugin runtime environment. I'm still not 100% sure and would appreciate any help.

**Reply by lukebakken** [reply]
Team RabbitMQ does not have the time or resources to figure out why our library does not work correctly in a "non official" Golang environment. This issue is best raised with Yaegi, if that is the environment you are using.

**Reply by roman-yerin** [reply]
Yeah, thanks.
FYI, the issue is fixed here https://github.com/killer-djon/rabbitmq-go which is 1 commit ahead

**Reply by lukebakken** [reply]
Please feel free to open a pull request and we will evaluate it.

Note that there ARE copy-paste bugs in the forked code that will result in errors!!!

For instance this line was clearly copied without noting that it should be `ch.receiveMethod`:

https://github.com/rabbitmq/amqp091-go/compare/main...killer-djon:rabbitmq-go:main#diff-7b28604b0113c10108ba3d8fc11fe87060c0e3bdeb770d9dfb0d3f4f9daa6edbR397

https://github.com/rabbitmq/amqp091-go/commit/7a97d245384a8d78c12de1852742c734fb3e71d9#commitcomment-135630795
